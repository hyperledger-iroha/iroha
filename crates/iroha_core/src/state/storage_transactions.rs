//! Multi-version append-only key value storage for transactions

#![allow(clippy::disallowed_types)]

use std::{borrow::Borrow, collections::HashSet, hash::Hash, num::NonZeroUsize, sync::Arc};

use arc_swap::ArcSwapOption;
use dashmap::DashMap;
use iroha_crypto::HashOf;
use iroha_data_model::prelude::SignedTransaction;
use parking_lot::{lock_api::MutexGuard, Mutex, RawMutex};
use serde::{Deserialize, Serialize};

type Key = HashOf<SignedTransaction>;
type Value = NonZeroUsize;

/// Multi-version append-only key value storage for transactions
/// This is analogue of [`mv::storage::Storage`] or `HashMap<Key, Value>`.
/// Contains hashes of transactions mapped onto block height where they are stored.
///
/// * Q: Why don't we use `HashMap`/`BTreeMap`?
///   A: Because we need multi-version storage with transactional behaviour
///   (`.view()`, `.block()`, `.block_and_revert()`, `.commit()`)
/// * Q: Why don't we use [`mv::storage::Storage`]?
///   A: Because transactions map consumes 80%+ RAM of iroha.
///   This storage is memory-optimized and consumes ~3x less memory.
pub struct TransactionsStorage {
    /// Latest block. Stored separately because of reverts.
    /// `None` when there are no blocks yet, otherwise must be not `None`.
    latest_block: ArcSwapOption<BlockInfo>,
    /// Map with aggregated transactions of multiple blocks, EXCEPT for the latest block.
    blocks: DashMap<Key, Value>,
    write_lock: Mutex<()>,
}

#[derive(Serialize, Deserialize)]
struct BlockInfo {
    /// Transactions added in the block
    transactions: HashSet<Key>,
    /// Height of the block.
    height: NonZeroUsize,
}

impl TransactionsStorage {
    /// Construct new [`Self`]
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self {
            latest_block: ArcSwapOption::empty(),
            blocks: DashMap::new(),
            write_lock: Mutex::new(()),
        }
    }

    /// Create persistent view of storage at certain point in time
    pub fn view(&self) -> TransactionsView {
        TransactionsView {
            latest_block: self.latest_block.load_full(),
            blocks: &self.blocks,
        }
    }

    /// Create block to aggregate updates
    pub fn block(&self) -> TransactionsBlock {
        self.block_impl(false)
    }

    /// Create block to aggregate updates and revert changes created in the latest block
    pub fn block_and_revert(&self) -> TransactionsBlock {
        self.block_impl(true)
    }

    fn block_impl(&self, revert: bool) -> TransactionsBlock {
        let guard = self.write_lock.lock();
        TransactionsBlock {
            latest_block_ref: &self.latest_block,
            blocks_ref: &self.blocks,
            _guard: guard,
            revert,
            current_block: None,
        }
    }
}

/// Persistent view of storage at certain point in time
pub trait TransactionsReadOnly {
    /// Read entry from the storage
    fn get<Q>(&self, key: &Q) -> Option<Value>
    where
        Key: Borrow<Q>,
        Q: Hash + Eq + ?Sized;
}

/// Module for [`TransactionsView`] and it's related impls
mod view {
    use super::*;

    /// Consistent view of the storage at the certain version
    #[derive(Serialize)]
    pub struct TransactionsView<'storage> {
        pub(super) latest_block: Option<Arc<BlockInfo>>,
        /// Some transactions may be added to the map after `Self` is created,
        /// but for us exists only transactions with `height < latest_block.height`
        pub(super) blocks: &'storage DashMap<Key, Value>,
    }

    impl TransactionsReadOnly for TransactionsView<'_> {
        fn get<Q>(&self, key: &Q) -> Option<Value>
        where
            Key: Borrow<Q>,
            Q: Hash + Eq + ?Sized,
        {
            let Some(block) = &self.latest_block else {
                return None;
            };
            let block_height = block.height;
            if block.transactions.contains(key) {
                return Some(block_height);
            }

            self.blocks
                .get(key)
                .map(|h| *h)
                .filter(|&h| h < block_height)
        }
    }
}
pub use view::TransactionsView;

/// Module for [`TransactionsBlock`] and it's related impls
mod block {
    use std::ops::Deref;

    use super::*;

    /// Batched update to the storage that can be reverted later
    pub struct TransactionsBlock<'storage> {
        /// References to [`TransactionsStorage`] struct
        pub(super) latest_block_ref: &'storage ArcSwapOption<BlockInfo>,
        pub(super) blocks_ref: &'storage DashMap<Key, Value>,
        pub(super) _guard: MutexGuard<'storage, RawMutex, ()>,

        /// Own fields
        pub(super) revert: bool,
        pub(super) current_block: Option<Arc<BlockInfo>>,
    }

    impl TransactionsBlock<'_> {
        /// Add transactions of the block
        pub fn insert_block(&mut self, transactions: HashSet<Key>, height: Value) {
            let block_info = BlockInfo {
                transactions,
                height,
            };
            assert!(
                self.current_block.is_none(),
                "`TransactionsBlock::insert_block()` must be called only once"
            );
            self.current_block = Some(Arc::new(block_info));
        }

        #[cfg(test)]
        pub fn insert_block_with_single_tx(&mut self, tx: Key, height: Value) {
            let transactions = [tx].into_iter().collect();
            self.insert_block(transactions, height);
        }

        /// Apply aggregated changes to the storage
        pub fn commit(self) {
            let previous_block = &self.latest_block_ref.load();
            let previous_block = previous_block.as_ref();

            let previous_height = previous_block.map_or(0, |b| b.height.get());
            let addition = usize::from(!self.revert);
            let expected_current_height = previous_height + addition;

            #[allow(clippy::option_if_let_else)]
            let current_block = match self.current_block {
                Some(current_block) => {
                    let current_height = current_block.height.get();
                    debug_assert_eq!(expected_current_height, current_height);
                    current_block
                }
                None => {
                    #[cfg(not(debug_assertions))]
                    panic!("`TransactionsBlock::insert_block()` was not called");

                    // Note that in production `current_block` must be not `None`.
                    // We support `None` case here for tests,
                    // in which `.block()` + `.commit()` is used to populate store,
                    // without actually adding block with transactions.
                    #[cfg(debug_assertions)]
                    Arc::new(BlockInfo {
                        transactions: HashSet::new(),
                        height: NonZeroUsize::new(expected_current_height).unwrap(),
                    })
                }
            };

            if !self.revert {
                if let Some(previous_block) = previous_block {
                    for &transaction in &previous_block.transactions {
                        self.blocks_ref.insert(transaction, previous_block.height);
                    }
                }
            }

            self.latest_block_ref.store(Some(current_block));
        }
    }

    impl TransactionsReadOnly for TransactionsBlock<'_> {
        fn get<Q>(&self, key: &Q) -> Option<Value>
        where
            Key: Borrow<Q>,
            Q: Hash + Eq + ?Sized,
        {
            if let Some(current_block) = &self.current_block {
                if current_block.transactions.contains(key) {
                    return Some(current_block.height);
                }
            }

            #[allow(clippy::explicit_deref_methods)]
            if let Some(latest_block) = self.latest_block_ref.load().deref() {
                if latest_block.transactions.contains(key) {
                    return Some(latest_block.height);
                }
            }

            self.blocks_ref.get(key).map(|h| *h)
        }
    }
}
pub use block::TransactionsBlock;

/// Module with serialization and deserialization of [`TransactionsStorage`]
mod serialization {
    use super::*;

    impl Serialize for TransactionsStorage {
        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: serde::Serializer,
        {
            let view = self.view();

            // Note that some new entries may be added to `blocks` during serialization.
            // We will filter such entries later during deserialization.
            // Potential alternative implementations:
            // * Clone `blocks`, filter, then serialize
            //   Bad approach, will have 2x peak memory usage
            // * Write custom serialization with filtration
            //   (using `serializer.serialize_map()`)
            view.serialize(serializer)
        }
    }

    impl<'de> Deserialize<'de> for TransactionsStorage {
        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            #[derive(Deserialize)]
            struct Storage {
                latest_block: Option<Arc<BlockInfo>>,
                blocks: DashMap<Key, Value>,
            }

            let storage = Storage::deserialize(deserializer)?;
            // See `serialize` implementation why filter is needed here
            if let Some(latest_block) = &storage.latest_block {
                let latest_block_height = latest_block.height;
                storage
                    .blocks
                    .retain(|_, &mut height| height < latest_block_height);
            } else {
                storage.blocks.clear();
            }

            Ok(TransactionsStorage {
                latest_block: ArcSwapOption::from(storage.latest_block),
                blocks: storage.blocks,
                write_lock: Mutex::new(()),
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn random_hash() -> Key {
        use rand::Rng;
        let bytes = rand::thread_rng().gen();
        let hash = iroha_crypto::Hash::prehashed(bytes);
        HashOf::from_untyped_unchecked(hash)
    }

    fn get_keys<const N: usize>() -> [Key; N] {
        [(); N].map(|()| random_hash())
    }

    fn get_values<const N: usize>() -> [Value; N] {
        let mut i = 0;
        [(); N].map(|()| {
            i += 1;
            NonZeroUsize::new(i).unwrap()
        })
    }

    fn insert_keys(block: &mut TransactionsBlock, keys: &[Key], value: Value) {
        let keys = keys.iter().copied().collect();
        block.insert_block(keys, value);
    }

    #[test]
    fn get() {
        let [k0, k1, k2, k3, k4] = get_keys();
        let [v1, v2, v3] = get_values();

        let storage = TransactionsStorage::new();
        let view0 = storage.view();

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k0, k1, k2], v1);
            block.commit()
        }
        let view1 = storage.view();

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k0, k1, k3], v2);
            block.commit()
        }
        let view2 = storage.view();

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k1, k4], v3);
            block.commit()
        }
        let view3 = storage.view();

        assert_eq!(view0.get(&k0), None);
        assert_eq!(view0.get(&k1), None);
        assert_eq!(view0.get(&k2), None);
        assert_eq!(view0.get(&k3), None);

        assert_eq!(view1.get(&k0), Some(v1));
        assert_eq!(view1.get(&k1), Some(v1));
        assert_eq!(view1.get(&k2), Some(v1));
        assert_eq!(view1.get(&k3), None);

        assert_eq!(view2.get(&k0), Some(v2));
        assert_eq!(view2.get(&k1), Some(v2));
        assert_eq!(view2.get(&k2), Some(v1));
        assert_eq!(view2.get(&k3), Some(v2));
        assert_eq!(view2.get(&k4), None);

        assert_eq!(view3.get(&k0), Some(v2));
        assert_eq!(view3.get(&k1), Some(v3));
        assert_eq!(view3.get(&k2), Some(v1));
        assert_eq!(view3.get(&k3), Some(v2));
        assert_eq!(view3.get(&k4), Some(v3));
    }

    #[test]
    fn revert() {
        let [k0] = get_keys();
        let [v1, v2] = get_values();

        let storage = TransactionsStorage::new();

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k0], v1);
            block.commit()
        }

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k0], v2);
            block.commit()
        }
        let view1 = storage.view();

        {
            let block = storage.block_and_revert();
            block.commit();
        }
        let view2 = storage.view();

        // View is persistent so revert is not visible
        assert_eq!(view1.get(&k0), Some(v2));
        // Revert is visible in the view created after revert was applied
        assert_eq!(view2.get(&k0), Some(v1));
    }

    #[test]
    fn serialization() {
        fn assert_views_equal(view1: &TransactionsView, view2: &TransactionsView, keys: &[Key]) {
            for key in keys {
                let value1 = view1.get(key);
                let value2 = view2.get(key);
                assert_eq!(value1, value2);
            }
        }
        fn check_view(view1: &TransactionsView, keys: &[Key]) {
            let json = serde_json::to_string(&view1).unwrap();
            let storage2: TransactionsStorage = serde_json::from_str(&json).unwrap();
            let view2 = storage2.view();
            assert_views_equal(view1, &view2, keys);
        }
        fn check_views(views: &[TransactionsView], keys: &[Key]) {
            for view in views {
                check_view(view, keys);
            }
        }

        let keys = get_keys();
        let [k0, k1, k2, k3, k4] = keys;
        let [v1, v2, v3] = get_values();

        let mut views = Vec::new();

        let storage = TransactionsStorage::new();
        views.push(storage.view());
        check_views(&views, &keys);

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k0, k1, k2], v1);
            block.commit()
        }
        views.push(storage.view());
        check_views(&views, &keys);

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k0, k1, k3], v2);
            block.commit()
        }
        views.push(storage.view());
        check_views(&views, &keys);

        {
            let mut block = storage.block();
            insert_keys(&mut block, &[k1, k4], v3);
            block.commit()
        }
        views.push(storage.view());
        check_views(&views, &keys);

        {
            let mut block = storage.block_and_revert();
            insert_keys(&mut block, &[k2], v3);
            block.commit()
        }
        views.push(storage.view());
        check_views(&views, &keys);
    }
}
