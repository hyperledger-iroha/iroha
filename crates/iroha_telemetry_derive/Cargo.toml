[package]
name = "iroha_telemetry_derive"

edition.workspace = true
version.workspace = true
authors.workspace = true

license.workspace = true

[lints]
workspace = true

[lib]
proc-macro = true

[badges]
is-it-maintained-issue-resolution = { repository = "https://github.com/hyperledger/iroha" }
is-it-maintained-open-issues = { repository = "https://github.com/hyperledger/iroha" }
maintenance = { status = "actively-developed" }

[features]
# FIXME: it doesn't work https://github.com/hyperledger/iroha/issues/5134
metric-instrumentation = []

[dependencies]
syn = { workspace = true }
quote = { workspace = true }
proc-macro2 = { workspace = true }
manyhow = { workspace = true }
iroha_macro_utils = { workspace = true }

[dev-dependencies]
iroha_core = { workspace = true }

trybuild = { workspace = true }
