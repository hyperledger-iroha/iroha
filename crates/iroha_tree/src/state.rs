//! Module for [`StateView`] and related components.

use super::*;

/// Represents the state view of each node.
pub type StateView = Tree<View>;

/// Each node value indicates the state view.
#[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
pub struct View;

impl Mode for View {
    type Authorizer = tr::AuthorizerV;
    type Parameter = tr::ParameterV;
    type Peer = tr::UnitV;
    type Domain = tr::DomainV;
    type Account = tr::UnitV;
    type Asset = tr::AssetV;
    type Nft = tr::NftV;
    type AccountAsset = tr::AccountAssetV;
    type Role = tr::UnitV;
    type Permission = tr::PermissionV;
    type AccountRole = tr::UnitV;
    type AccountPermission = tr::UnitV;
    type RolePermission = tr::UnitV;
    type Trigger = tr::TriggerV;
    type Condition = tr::ConditionV;
    type Executable = tr::ExecutableV;
    type TriggerCondition = tr::UnitV;
    type TriggerExecutable = tr::UnitV;
    type DomainMetadata = tr::MetadataV;
    type AccountMetadata = tr::MetadataV;
    type AssetMetadata = tr::MetadataV;
    type NftData = tr::MetadataV;
    type TriggerMetadata = tr::MetadataV;
    type DomainAdmin = tr::UnitV;
    type AssetAdmin = tr::UnitV;
    type NftAdmin = tr::UnitV;
    type NftOwner = tr::UnitV;
    type TriggerAdmin = tr::UnitV;
}

impl NodeReadWrite for StateView {
    type Status = event::Event;

    fn as_status(&self) -> Self::Status {
        self.iter()
            .map(|(k, v)| NodeEntry::try_from((k.clone(), v.into())).unwrap())
            .collect()
    }
}

/// Interface for interacting with the main state of the application.
pub trait WorldState {
    /// Indicates that the write request was rejected due to data integrity violations.
    type InvariantViolation: From<Box<NodeConflict<changeset::Write>>>;

    /// Applies a write entry to the state.
    ///
    /// # Errors
    ///
    /// Fails if the update violates data integrity constraints.
    fn update_by(
        &mut self,
        entry: NodeEntry<changeset::Write>,
    ) -> Result<(), Self::InvariantViolation>;

    /// Scans for inconsistencies based on event predictions and attempts to resolve them, returning an additional changeset.
    ///
    /// # Errors
    ///
    /// Fails if the expected change is determined to break data integrity.
    fn sanitize(
        &self,
        event_prediction: &event::Event,
    ) -> Result<changeset::ChangeSet, Self::InvariantViolation>;

    /// Retrieve stored values based on the `readset` query.
    fn load(&self, readset: &readset::ReadSet) -> StateView;

    /// Applies an unordered changeset to the state, resulting in events.
    ///
    /// # Errors
    ///
    /// Fails if the update violates data integrity constraints.
    fn update(
        &mut self,
        changeset: changeset::ChangeSet,
    ) -> Result<event::Event, Self::InvariantViolation> {
        let event_prediction = changeset.as_status();
        let changeset = (changeset + self.sanitize(&event_prediction)?)?;
        let event = changeset.as_status();

        for (key, value) in changeset {
            self.update_by(NodeEntry::try_from((key, value)).unwrap())?;
        }

        Ok(event)
    }
}

#[allow(missing_docs)]
pub mod transitional {
    use hashbrown::{HashMap, HashSet};

    use super::*;

    /// State view at `Unit` type nodes.
    #[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
    pub struct UnitV;

    /// State view at `Authorizer` type nodes.
    #[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
    pub struct AuthorizerV;

    /// State view at `Parameter` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub struct ParameterV {
        pub(crate) parameter: dm::Parameter,
    }

    /// State view at `Domain` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub struct DomainV {
        pub(crate) logo: Option<dm::IpfsPath>,
    }

    /// State view at `Asset` type nodes.
    #[derive(Debug, PartialEq, Eq, Constructor, Clone, Decode, Encode)]
    pub struct AssetV {
        pub(crate) total_quantity: dm::Numeric,
        pub(crate) mintable: dm::Mintable,
        pub(crate) logo: Option<dm::IpfsPath>,
    }

    /// State view at `Nft` type nodes.
    #[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
    pub struct NftV;

    /// State view at `AccountAsset` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub struct AccountAssetV {
        pub(crate) balance: dm::Numeric,
    }

    /// State view at `Permission` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub struct PermissionV {
        pub(crate) permission: permission::Permission,
    }

    /// State view at `Trigger` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub struct TriggerV {
        pub(crate) repeats: dm::Repeats,
    }

    /// State view at `Condition` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub enum ConditionV {
        World(receptor::Receptor),
        Time(dm::TimeSchedule),
        Block(BlockCommit),
    }

    #[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
    pub struct BlockCommit;

    /// State view at `Executable` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub enum ExecutableV {
        Static(changeset::ChangeSet),
        Dynamic(WasmExecutable),
    }

    #[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
    pub struct WasmExecutable;

    /// State view at `Metadata` type nodes.
    #[derive(Debug, PartialEq, Eq, From, Clone, Decode, Encode)]
    pub struct MetadataV {
        pub(crate) json: dm::Json,
    }

    #[derive(Debug, PartialEq, Eq, Constructor, Clone)]
    pub struct TriggerEntry<'a> {
        pub id: &'a dm::TriggerId,
        pub condition: &'a ConditionV,
        pub executable: &'a ExecutableV,
    }

    #[derive(Debug, PartialEq, Eq, Constructor, Clone)]
    pub struct WorldTriggerEntry<'a> {
        pub id: &'a dm::TriggerId,
        pub receptor: &'a receptor::Receptor,
        pub executable: &'a ExecutableV,
    }

    impl<'a> TryFrom<TriggerEntry<'a>> for WorldTriggerEntry<'a> {
        type Error = &'static str;

        fn try_from(entry: TriggerEntry<'a>) -> Result<Self, Self::Error> {
            match entry.condition {
                ConditionV::World(receptor) => {
                    Ok(WorldTriggerEntry::new(entry.id, receptor, entry.executable))
                }
                _ => Err("conversion succeeds only when this trigger subscribes to world events"),
            }
        }
    }

    impl StateView {
        pub fn triggers(&self) -> impl Iterator<Item = TriggerEntry> {
            let ids = self.keys().filter_map(|k| match k {
                NodeKey::Trigger(id) => Some(&**id),
                _ => None,
            });
            let conditions: HashMap<_, _> = self
                .keys()
                .filter_map(|k| match k {
                    NodeKey::TriggerCondition((trg, con)) => {
                        let Some(NodeValue::Condition(condition)) =
                            self.get(&NodeKey::Condition(con.clone()))
                        else {
                            panic!("should be loaded into the state view")
                        };
                        Some((&**trg, condition))
                    }
                    _ => None,
                })
                .collect();
            let executables: HashMap<_, _> = self
                .keys()
                .filter_map(|k| match k {
                    NodeKey::TriggerExecutable((trg, exe)) => {
                        let Some(NodeValue::Executable(executable)) =
                            self.get(&NodeKey::Executable(exe.clone()))
                        else {
                            panic!("should be loaded into the state view")
                        };
                        Some((&**trg, executable))
                    }
                    _ => None,
                })
                .collect();

            ids.map(move |id| {
                let condition = *conditions
                    .get(id)
                    .expect("should be loaded into the state view");
                let executable = *executables
                    .get(id)
                    .expect("should be loaded into the state view");
                TriggerEntry::new(id, condition, executable)
            })
        }

        pub fn world_triggers(&self) -> impl Iterator<Item = WorldTriggerEntry> {
            self.triggers().filter_map(|entry| entry.try_into().ok())
        }
    }

    impl TriggerEntry<'_> {
        pub fn leads_to_event_loop(&self, state: &StateView) -> bool {
            let empty_receptor = receptor::Receptor::default();
            let mut world_triggers: HashMap<_, _> = state
                .world_triggers()
                .map(|entry| (entry.id, (entry.receptor, entry.executable)))
                .collect();
            let receptor = match self.condition {
                ConditionV::World(receptor) => receptor,
                _ => &empty_receptor,
            };
            world_triggers.insert(self.id, (receptor, self.executable));
            let mut stack = vec![self.id];
            let mut seen = HashSet::new();
            while let Some(trigger_id) = stack.pop() {
                if seen.contains(&trigger_id) {
                    return true;
                }
                seen.insert(trigger_id);
                let event_expected = match &world_triggers[trigger_id].1 {
                    state::tr::ExecutableV::Static(changeset) => changeset.as_status(),
                    state::tr::ExecutableV::Dynamic(_wasm) => {
                        todo!("Wasm executable should declare the union of possible events (#5362)")
                    }
                };
                if event_expected.iter().any(|(_k, v)| {
                    matches!(
                        v,
                        NodeValue::TriggerCondition(event::UnitS::Create)
                            | NodeValue::TriggerExecutable(event::UnitS::Create)
                    )
                }) {
                    // Trigger registration by another trigger is not allowed unless Wasm executables declare the candidate trigger executables.
                    return true;
                }
                let next_trigger_ids = world_triggers.iter().filter_map(|(id, (receptor, _))| {
                    event_expected.passes(receptor).is_ok().then_some(id)
                });
                stack.extend(next_trigger_ids);
            }
            false
        }
    }

    impl TryFrom<(dm::AccountId, dm::Executable)> for ExecutableV {
        type Error = Box<NodeConflict<changeset::Write>>;

        fn try_from(
            (authority, executable): (dm::AccountId, dm::Executable),
        ) -> Result<Self, Self::Error> {
            match executable {
                dm::Executable::Instructions(instructions) => {
                    let changeset =
                        changeset::ChangeSet::try_from((authority, instructions.into_vec()))?;
                    Ok(changeset.into())
                }
                dm::Executable::Wasm(_wasm) => Ok(WasmExecutable.into()),
            }
        }
    }
}

pub use transitional as tr;

#[cfg(test)]
mod tests {
    #[cfg(not(feature = "std"))]
    use alloc::format;

    use dm::{DomainId, Repeats, TriggerId};

    use super::{transitional::TriggerEntry, *};
    use crate::{
        changeset::{ChangeSet, ConditionW, DomainW, ExecutableW, TriggerW, UnitW},
        receptor::Receptor,
    };

    /// See the corresponding integration test `triggers::not_registered_when_potential_event_loop_detected`.
    #[expect(clippy::too_many_lines)]
    #[test]
    fn detects_event_loop() {
        // Subscribes to changes in the domain "dom_{i}" with statuses "{s}".
        let receptor = |i: usize, s: &str| {
            Receptor::from_iter([fuzzy_node!(
                Domain,
                some!(DomainId::from_str(&format!("dom_{i}")).unwrap()),
                FilterU8::from_str(s).unwrap()
            )])
        };
        // Publishes the deletion of the domain "dom_{j}".
        let changeset = |j: usize| {
            ChangeSet::from_iter([node!(
                Domain,
                DomainId::from_str(&format!("dom_{j}")).unwrap(),
                DomainW::Delete(())
            )])
        };
        // Bridges the above subscriber and publisher.
        let trigger = |i: usize, s: &str, j: usize| {
            let condition = tr::ConditionV::from(receptor(i, s));
            let executable = tr::ExecutableV::from(changeset(j));
            (
                (
                    TriggerId::from_str(&format!("trg_{i}{s}_{j}d")).unwrap(),
                    tr::TriggerV::from(Repeats::Indefinitely),
                ),
                (crate::tr::ConditionId::from(&condition), condition),
                (crate::tr::ExecutableId::from(&executable), executable),
            )
        };
        // A potential connection exists through the deletion of "dom_1".
        let (trg_0d_1d, trg_1d_2d) = (trigger(0, "d", 1), trigger(1, "d", 2));
        // The state after registering the above triggers.
        let state = StateView::from_iter([
            node!(Condition, trg_0d_1d.1 .0.clone(), trg_0d_1d.1 .1),
            node!(Condition, trg_1d_2d.1 .0.clone(), trg_1d_2d.1 .1),
            node!(Executable, trg_0d_1d.2 .0.clone(), trg_0d_1d.2 .1),
            node!(Executable, trg_1d_2d.2 .0.clone(), trg_1d_2d.2 .1),
            node!(
                TriggerCondition,
                trg_0d_1d.0 .0.clone(),
                trg_0d_1d.1 .0,
                tr::UnitV
            ),
            node!(
                TriggerCondition,
                trg_1d_2d.0 .0.clone(),
                trg_1d_2d.1 .0,
                tr::UnitV
            ),
            node!(
                TriggerExecutable,
                trg_0d_1d.0 .0.clone(),
                trg_0d_1d.2 .0,
                tr::UnitV
            ),
            node!(
                TriggerExecutable,
                trg_1d_2d.0 .0.clone(),
                trg_1d_2d.2 .0,
                tr::UnitV
            ),
            node!(Trigger, trg_0d_1d.0 .0, trg_0d_1d.0 .1),
            node!(Trigger, trg_1d_2d.0 .0, trg_1d_2d.0 .1),
        ]);

        for (entry, leads_to_event_loop) in [
            // Short-circuiting.
            (trigger(2, "d", 0), true),
            // No short-circuiting due to status mismatch.
            (trigger(2, "cu", 0), false),
            // Extending the graph.
            (trigger(2, "d", 3), false),
            // Creating another cyclic cluster.
            (trigger(3, "d", 3), true),
            // Creating another acyclic cluster.
            (trigger(3, "d", 4), false),
            {
                let mut trg_3d_x = trigger(3, "d", 4);
                let another = trigger(10, "", 20);
                trg_3d_x.2 .1 = ChangeSet::from_iter([
                    node!(
                        Condition,
                        another.1 .0.clone(),
                        ConditionW::Set(another.1 .1)
                    ),
                    node!(
                        Executable,
                        another.2 .0.clone(),
                        ExecutableW::Set(another.2 .1)
                    ),
                    node!(
                        TriggerCondition,
                        another.0 .0.clone(),
                        another.1 .0,
                        UnitW::Create(())
                    ),
                    node!(
                        TriggerExecutable,
                        another.0 .0.clone(),
                        another.2 .0,
                        UnitW::Create(())
                    ),
                    node!(Trigger, another.0 .0, TriggerW::Create(another.0 .1)),
                ])
                .into();
                // Creating an additional trigger.
                (trg_3d_x, true)
            },
        ]
        .iter()
        .map(|(trg, b)| (TriggerEntry::new(&trg.0 .0, &trg.1 .1, &trg.2 .1), *b))
        {
            assert_eq!(leads_to_event_loop, entry.leads_to_event_loop(&state));
        }
    }

    #[test]
    fn passes_permission() {
        use permission::Permission;

        let key = |i: usize| dm::RoleId::from_str(&format!("role_{i}")).unwrap();
        let views = [
            StateView::default(),
            StateView::from_iter([node!(Role, key(0), tr::UnitV)]),
            StateView::from_iter([
                node!(Role, key(0), tr::UnitV),
                node!(Role, key(1), tr::UnitV),
            ]),
        ];
        let permissions = [
            Permission::default(),
            Permission::from_iter([fuzzy_node!(Role, some!(key(0)), event::UnitS::Read)]),
            Permission::from_iter([fuzzy_node!(Role, None, FilterU8::from_str("r").unwrap())]),
        ];

        let missing_permission = views[2].passes(&permissions[1]).unwrap_err();
        let complemented_permission = permissions[1].clone() | missing_permission;
        assert!(views[2].passes(&complemented_permission).is_ok());

        for (i, view) in views.iter().enumerate() {
            for (j, permission) in permissions.iter().enumerate() {
                assert_eq!(i <= j, view.passes(permission).is_ok());
            }
        }
    }
}
