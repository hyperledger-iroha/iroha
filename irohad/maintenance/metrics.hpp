/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#ifndef IROHA_MAINTENANCE_METRICS_HPP
#define IROHA_MAINTENANCE_METRICS_HPP

#include <string>
#include <memory>
#include <prometheus/registry.h>

std::shared_ptr<prometheus::Registry>
    maintenance_metrics_init(std::string const& listen_addr);

#endif //IROHA_MAINTENANCE_METRICS_HPP
