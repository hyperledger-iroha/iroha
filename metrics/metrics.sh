#!/usr/bin/env bash
set -euo pipefail

cd $(dirname $(realpath ${BASH_SOURCE[0]}))
prometheus --config.file=prometheus.yml
