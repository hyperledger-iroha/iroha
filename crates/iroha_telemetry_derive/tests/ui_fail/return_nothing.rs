use iroha_core::state::StateTransaction;
use iroha_telemetry_derive::metrics;

#[metrics(+"test_query", "another_test_query_without_timing")]
fn execute(_state_transaction: &StateTransaction) {
    Ok::<(), ()>(());
}

fn main() {}
