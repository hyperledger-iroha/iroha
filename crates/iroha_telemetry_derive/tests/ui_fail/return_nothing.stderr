error: `Fn` must return `Result`. Returns nothing instead.
 --> tests/ui_fail/return_nothing.rs:5:1
  |
5 | #[metrics(+"test_query", "another_test_query_without_timing")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `metrics` (in Nightly builds, run with -Z macro-backtrace for more info)
